### ISSUE TITLE
Plan for unix support

### ISSUE THREAD

**Issue by david-bouyssie** [OpenByteDev/netcorehost#2]
Do you have planned to add support for unix platforms?

According to this example it seems to be doable:
https://github.com/dotnet/samples/blob/main/core/hosting/HostWithHostFxr/src/NativeHost/nativehost.cpp

However I have noticed a distinction regarding strings which seem to be wide on windows (```typedef wchar_t char_t```) but not on unix (```typedef char char_t```).

Would be happy to contribute to this addition but help might be needed.

**Comment by OpenByteDev** [comment]
I plan to add unix support, but it is not high on my priority list as I dont personally use it.

**Comment by david-bouyssie** [comment]
Ok great, let me know if I can help on PR.

**Comment by david-bouyssie** [comment]
Well, if you give me a hint of what I should change first, I can work on it.
I have already identified the WideString vs CString difference, but I may have missed other points.

**Comment by OpenByteDev** [comment]
The difference in string types should be the only change required.

**Comment by OpenByteDev** [comment]
I created an initial version that supports os other than windows.
It builds, but there are some weird errors on the tests: https://github.com/OpenByteDev/netcorehost/runs/2180322999?check_suite_focus=true

Do you have any idea about the problem?

**Comment by david-bouyssie** [comment]
Wow awesome, thank you for this effort!

It seems that you are linking C++ code with rust. So you have to include the C++ standard library.
Can you please try with "cargo:rustc-link-lib=dylib=stdc++"?

For reference: https://users.rust-lang.org/t/linking-c-on-rust-undefined-reference-to-operator-delete-void-and-gxx-personality-v0-libstdc-missing/49658

**Comment by OpenByteDev** [comment]
I discovered this yesterday, but the tests are still failing (but with less errors).
Interestingly they only seem to fail when run together...

**Comment by OpenByteDev** [comment]
To clarify:
- Building succeeds without error.
- When a specific test is run with `cargo test --test [testname]` (e.g. `cargo test --test hello_world`) every test succeeds.
- When all tests are run at once with `cargo test`  linking fails with this error:
```
/opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/iostream:74: undefined reference to `std::ios_base::Init::Init()'
```
The full error log is available here: https://github.com/OpenByteDev/netcorehost/runs/2192279154?check_suite_focus=true

**Comment by david-bouyssie** [comment]
This is weird, in this contexts some headers seem to be loaded from a non-existing directory:
```/opt/rh/devtoolset-7/root/usr/```

I can reproduce the issue on Ubuntu 18.04 (WSL)

**Comment by david-bouyssie** [comment]
In bindings\nethost.rs I added this:
```
#[link(name = "nethost", kind = "static")]
#[link(name = "stdc++", kind = "dylib")]
extern "C" {
    /// Get the path to the hostfxr library
...
```

And I removed from build.rs:
```
println!("cargo:rustc-link-lib=dylib=stdc++");
println!("cargo:rustc-link-lib=static=nethost");
```

This seem to work.

**Comment by OpenByteDev** [comment]
Nice!
Do you know why this works and my version did not?

**Comment by david-bouyssie** [comment]
No, sorry, I'm missing knowledge there. But maybe on the rust gitter you may have a chance to get some clues.
I'm not even sure if the proposed solution is the way to go.

**Comment by david-bouyssie** [comment]
BTW, I found this other interesting project:
https://github.com/AaronRobinsonMSFT/DNNE

It's also based on nethost and provides an utility to generate C source files to be linked against nethost (from exported functions with or without UnmanagedCallersOnly).

**Comment by OpenByteDev** [comment]
It seems that the `link` attribute does not work for windows? (at least in this situation; again linking errors)

**Comment by david-bouyssie** [comment]
I have seen your fix in the commit https://github.com/OpenByteDev/netcorehost/commit/0afdcddd9ea3c03dbbbaa5823d1cd73932082c2d

Looks good to me :)

Thanks!!!

**Comment by OpenByteDev** [comment]
The linking weirdness seems to have been fixed in 1.61 (probably due to https://github.com/rust-lang/rust/pull/93901/)
